src/exoforce.py

```diff
@@ -13,147 +19,116 @@
-print("\n")
-
-jointinfo = p.getJointInfo(human, hand_link)
-print("The joint info or link hand_link is: ", jointinfo)
-
-# NUMBER OF BODIES = 2 (0 = plane, 1 = human)
-bodies = p.getNumBodies()
-print("The number of bodies is: ", bodies)
-"""for i in range(bodies):
-	info = p.getBodyInfo(i)
-	body_id = p.getBodyUniqueId(i)
-	print("Info: ",info)
-	print("This body's Id is: ", body_id)"""
-
 
-# COMMON SETUP
+class humanFigure():
+	def __init__(self, body_id):
+		self.body_id = body_id
+
+	def find_link_number(self, link_name):
+		numJoints = p.getNumJoints(self.body_id)
+		link = None
+		for i in range(numJoints):
+			info = p.getJointInfo(self.body_id,i)
+			if info[12] == link_name:
+				link = i;
+		return link
+
+
+class Tendon():
+	def __init__(self, body_id, link, end_displacement, i):
+		self.body_id = body_id
+		self.link = link
+		self.i = i
+		self.end_displacement = end_displacement
+		self.forceId = p.addUserDebugParameter(str("tendon " + str(self.i)), 0,200,0)
+		self.start_location = p.getLinkState(self.body_id, self.link)[0]
+		self.end_location = [self.start_location[0]+self.end_displacement[0],
+				     self.start_location[1]+self.end_displacement[1],
+				     self.start_location[2]+self.end_displacement[2]]
+		self.tendon = p.addUserDebugLine(self.start_location,
+		                      		 self.end_location,
+                                      		 lineColorRGB=[0,0,255],
+                                      		 lineWidth=2)
+
+	def decompose_force(self):
+		motor_force = p.readUserDebugParameter(self.forceId)
+		force_vector = [a_i - b_i for a_i, b_i in zip(self.end_location,self.start_location)]
+		self.force_components = [motor_force * x/norm(force_vector) for x in force_vector]
+
+	def apply_force(self):
+		p.applyExternalForce(objectUniqueId=self.body_id,
+			     	     linkIndex = self.link,
+			     	     forceObj = self.force_components,
+			     	     posObj=self.start_location,
+			     	     flags=p.WORLD_FRAME)
+
+	def update(self):
+		self.start_location = p.getLinkState(self.body_id, self.link)[0]
+		p.addUserDebugLine(self.start_location,self.end_location,lineColorRGB=[0,0,255],
+				   lineWidth=2, replaceItemUniqueId=self.tendon)
+		self.decompose_force()
+		self.apply_force()
+		
+#######################################################################################
+###############                                                         ###############
+###############               LINK, WORLD AND TENDON SETUP              ###############
+###############                                                         ###############
+#######################################################################################
+
+
+# LINK DEFINITIONS
+hf = humanFigure(human)
+
+lefthand_link = hf.find_link_number(b'human/left_wrist')
+righthand_link = hf.find_link_number(b'human/right_wrist')
+
+upperchest_link = hf.find_link_number(b'human/shoulder_center')
```

Comment:
also unclear why do you define all these link ids that are never used

Reply:
When we are done with the optimization of key link points we will probably add more tendons on other links and so, I thought that preparing them might be helpful moving forward. This doesn't apply to legs though. I will delete those straight away. Should I also delete the other unused ones?